rir/src/interpreter/interp.cpp

```diff
@@ -1232,43 +1232,54 @@ static SEXP seq_int(int n1, int n2) {
     return ans;
 }
 
-#define BINDING_CACHE_SIZE 5
+#define BINDING_CACHE_SIZE 8
+// must be a power of 2 for modulus using & CACHE_MASK to work
+#define CACHE_MASK (BINDING_CACHE_SIZE - 1)
+
 typedef struct {
     SEXP loc;
     Immediate idx;
 } BindingCache;
 
-static RIR_INLINE SEXP cachedGetBindingCell(SEXP env, Immediate idx,
+static RIR_INLINE SEXP cachedGetBindingCell(SEXP env, Immediate poolIdx,
+                                            Immediate cacheIdx,
                                             InterpreterInstance* ctx,
-                                            BindingCache* bindingCache) {
+                                            BindingCache* bindingCache,
+                                            bool smallCache) {
     if (env == R_BaseEnv || env == R_BaseNamespace)
         return NULL;
 
-    Immediate cidx = idx % BINDING_CACHE_SIZE;
-    if (bindingCache[cidx].idx == idx) {
+    Immediate cidx;
+    if (smallCache)
+        cidx = cacheIdx;
+    else
+        cidx = cacheIdx & CACHE_MASK;
+
+    if (bindingCache[cidx].idx == cacheIdx)
```

Comment:
i would go for
```
bool hit = smallCache || bindingCache[cidx].idx == cacheIdx;
if (hit && bindingCache[cidx].loc)
  ...
```

this should be a bit faster for small cache and additionally solve the zero index issue

Reply:
`cacheLocation - 1` implies substracting on every access, I'd prefer to just not use the zero index, its only 16 bytes lost. 

Testing for `loc!=null` enables to use the zero index too, but it implies to add one more check for the nonSmall version (cacheIdx + loc != null) while it adds nothing to the smallcache (previously testing for idx now for loc)